### PR TITLE
feat: inactive accounts revocation

### PR DESCRIPTION
**Pull request by rahmatrhd**
### Changes:
- New fields for Grant: `RequestedExpirationDate` and `ExpirationDateReason`
- Refactor job handler config
- New job handler: `revokeGrantsByUserCriteria`

### CODE REVIEW

File: jobs/revoke_grants_by_user_criteria.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+	"github.com/goto/guardian/plugins/identities"
+)
+
+type RevokeGrantsByUserCriteriaConfig struct {
+	IAM                 domain.IAMConfig     `mapstructure:"iam"`
+	UserCriteria        evaluator.Expression `mapstructure:"user_criteria"`
+	ReassignOwnershipTo evaluator.Expression `mapstructure:"reassign_ownership_to"`
+}
+
+func (h *handler) RevokeGrantsByUserCriteria(ctx context.Context, c Config) error {
```
Comment (bsushmith) [inline jobs/revoke_grants_by_user_criteria.go:18]: Can you add some debug level logs also in this method?

File: domain/grant.go [hunk 2]
```diff
@@ -22,25 +22,27 @@ const (
...
+	StatusInProvider        GrantStatus `json:"status_in_provider" yaml:"status_in_provider"`
+	AccountID               string      `json:"account_id" yaml:"account_id"`
+	AccountType             string      `json:"account_type" yaml:"account_type"`
+	ResourceID              string      `json:"resource_id" yaml:"resource_id"`
+	Role                    string      `json:"role" yaml:"role"`
+	Permissions             []string    `json:"permissions" yaml:"permissions"`
+	IsPermanent             bool        `json:"is_permanent" yaml:"is_permanent"`
+	ExpirationDate          *time.Time  `json:"expiration_date" yaml:"expiration_date"`
+	RequestedExpirationDate *time.Time  `json:"requested_expiration_date,omitempty" yaml:"requested_expiration_date,omitempty"`
+	ExpirationDateReason    string      `json:"expiration_date_reason,omitempty" yaml:"expiration_date_reason,omitempty"`
```
Comment (mabdh) [inline domain/grant.go:36]: Why not just

```suggestion
	ExpirationReason    string      `json:"expiration_reason,omitempty" yaml:"expiration_reason,omitempty"`
```

?

File: jobs/jobs.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+const (
+	FetchResources             Type = "fetch_resources"
+	ExpiringGrantNotification  Type = "expiring_grant_notification"
+	RevokeExpiredGrants        Type = "revoke_expired_grants"
+	RevokeGrantsByUserCriteria Type = "revoke_grants_by_user_criteria"
+
+	// Deprecated: use RevokeExpiredGrants instead
+	RevokeExpiredAccess Type = "revoke_expired_access"
+	// Deprecated: use ExpiringGrantNotification instead
+	ExpiringAccessNotification Type = "expiring_access_notification"
```
Comment (mabdh) [inline jobs/jobs.go:16]: To make it clear it is a type constant, maybe we could prepend with `Type...` ? Would be easier to navigate with autocomplete too.

```suggestion
	TypeFetchResources             Type = "fetch_resources"
	TypeExpiringGrantNotification  Type = "expiring_grant_notification"
	TypeRevokeExpiredGrants        Type = "revoke_expired_grants"
	TypeRevokeGrantsByUserCriteria Type = "revoke_grants_by_user_criteria"

	// Deprecated: use RevokeExpiredGrants instead
	TypeRevokeExpiredAccess Type = "revoke_expired_access"
	// Deprecated: use ExpiringGrantNotification instead
	TypeExpiringAccessNotification Type = "expiring_access_notification"
```

File: domain/grant.go (hunk 2 above)
Comment (rahmatrhd) [reply domain/grant.go:36]: initially did this, but I think it could be misunderstood as the reason why the grant got expired, wdyt?

Comment (mabdh) [reply domain/grant.go:36]: I think it is okay to use `ExpirationDateReason` for this then.

File: jobs/revoke_grants_by_user_criteria.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+		} else if !criteriaSatisfied {
+			continue
+		}
+
+		// revoking grants with account_id == email
+		revokeGrantsFilter := domain.RevokeGrantsFilter{
+			AccountIDs: []string{email},
+		}
+		h.logger.Info("revoking grants", "account_id", email)
+		// revokedGrants, err := h.grantService.BulkRevoke(ctx, revokeGrantsFilter, domain.SystemActorName, "Revoked due to user deactivated")
```
Comment (mabdh) [inline jobs/revoke_grants_by_user_criteria.go]: maybe forgot to remove this?

Comment (rahmatrhd) [reply jobs/revoke_grants_by_user_criteria.go]: @mabdh it has been updated, please check the latest changes

**Review by mabdh — APPROVED** [review]

**Comment by bsushmith** [comment]
Can you update this message also - https://github.com/goto/guardian/pull/27#issue-1693376480

**Pull request merged**